File: sdk/contextmesh.js
/**
 * ContextMesh JavaScript/TypeScript SDK
 * npm install contextmesh
 *
 * Quick start:
 *   import { Mesh } from "contextmesh";
 *   const mesh = new Mesh("cm_live_your_key");
 *   await mesh.remember("prod DB is postgres 15 on AWS us-east-1");
 *   const results = await mesh.query("what do we know about our database?");
 */

const DEFAULT_BASE = "https://api.contextmesh.dev";

export class ContextMeshError extends Error {
  constructor(message, statusCode) {
    super(message);
    this.name = "ContextMeshError";
    this.statusCode = statusCode;
  }
}

export class Mesh {
  constructor(apiKey, baseUrl) {
    if (!apiKey) throw new ContextMeshError("apiKey is required. Get one at https://contextmesh.dev", 0);
    this.key     = apiKey;
    this.base    = (baseUrl || DEFAULT_BASE).replace(/\/$/, "");
    this.headers = {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    };
  }

  async _fetch(method, path, body, params) {
    let url = `${this.base}${path}`;
    if (params) {
      const qs = new URLSearchParams(params).toString();
      if (qs) url += `?${qs}`;
    }
    const res = await fetch(url, {
      method,
      headers: this.headers,
      body: body ? JSON.stringify(body) : undefined,
    });
    if (res.status === 401) throw new ContextMeshError("Invalid API key.", 401);
    if (res.status === 429) {
      const data = await res.json().catch(() => ({}));
      throw new ContextMeshError(`Rate/quota limit exceeded. ${data.detail ?? ""}`, 429);
    }
    if (!res.ok) {
      const text = await res.text();
      throw new ContextMeshError(`API error ${res.status}: ${text}`, res.status);
    }
    return res.json();
  }

  /**
   * Store context. Returns the entry ID.
   * @example
   * const id = await mesh.remember("API rate limit is 1000 req/min");
   * const id = await mesh.remember("Use camelCase for JS", { tags: ["conventions"] });
   */
  async remember(text, opts = {}) {
    const result = await this._fetch("POST", "/remember", {
      text,
      tags:         opts.tags        ?? [],
      source_agent: opts.sourceAgent ?? null,
      confidence:   opts.confidence  ?? 1.0,
    });
    return result.id;
  }

  /**
   * Semantic search. Returns ranked list of context entries.
   * @example
   * const hits = await mesh.query("what do we know about our database?");
   * hits.forEach(h => console.log(h.text, h.score));
   */
  async query(q, opts = {}) {
    const result = await this._fetch("POST", "/query", {
      q,
      top_k:      opts.topK     ?? 5,
      tag_filter: opts.tag      ?? null,
      min_score:  opts.minScore ?? 0.3,
    });
    return result.results;
  }

  /**
   * Delete a context entry by ID.
   */
  async forget(entryId) {
    await this._fetch("DELETE", `/forget/${entryId}`);
    return true;
  }

  /**
   * Browse stored entries.
   */
  async list(opts = {}) {
    const params = {
      limit:  String(opts.limit  ?? 50),
      offset: String(opts.offset ?? 0),
    };
    if (opts.tag) params.tag_filter = opts.tag;
    return this._fetch("GET", "/list", undefined, params);
  }

  /**
   * Check plan usage for this month.
   */
  async usage() {
    return this._fetch("GET", "/usage");
  }

  /**
   * Bulk store multiple items.
   */
  async rememberMany(items) {
    return Promise.all(
      items.map(item =>
        typeof item === "string" ? this.remember(item) : this.remember(item.text, item)
      )
    );
  }
}

// CommonJS compat
module.exports = { Mesh, ContextMeshError };
module.exports.default = { Mesh, ContextMeshError };

File: sdk/contextmesh.py
"""
ContextMesh Python SDK
pip install contextmesh

Quick start:
    from contextmesh import Mesh
    mesh = Mesh("cm_live_your_key")
    mesh.remember("prod DB is postgres 15 on AWS us-east-1")
    results = mesh.query("what should I know about our database?")
"""

import httpx
import asyncio
from typing import Optional, Union


DEFAULT_BASE = "https://contextmesh.dev"


class ContextMeshError(Exception):
    pass


class Mesh:
    """
    Synchronous client. Works everywhere.
    For async usage, use AsyncMesh.
    """

    def __init__(self, api_key: str, base_url: str = DEFAULT_BASE):
        if not api_key:
            raise ContextMeshError("api_key is required. Get one at https://contextmesh.dev")
        self._key     = api_key
        self._base    = base_url.rstrip("/")
        self._headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type":  "application/json",
        }

    def _post(self, path: str, body: dict) -> dict:
        with httpx.Client(timeout=15.0) as c:
            r = c.post(f"{self._base}{path}", headers=self._headers, json=body)
            self._raise(r)
            return r.json()

    def _get(self, path: str, params: dict = None) -> dict:
        with httpx.Client(timeout=15.0) as c:
            r = c.get(f"{self._base}{path}", headers=self._headers, params=params or {})
            self._raise(r)
            return r.json()

    def _delete(self, path: str) -> dict:
        with httpx.Client(timeout=15.0) as c:
            r = c.delete(f"{self._base}{path}", headers=self._headers)
            self._raise(r)
            return r.json()

    def _raise(self, r: httpx.Response):
        if r.status_code == 401:
            raise ContextMeshError("Invalid API key.")
        if r.status_code == 429:
            raise ContextMeshError(f"Rate/quota limit exceeded. {r.json().get('detail','')}")
        if r.status_code >= 400:
            raise ContextMeshError(f"API error {r.status_code}: {r.text}")

    def remember(
        self,
        text: str,
        tags: list[str] = None,
        source_agent: str = None,
        confidence: float = 1.0,
    ) -> str:
        """
        Store context. Returns the entry ID.

        mesh.remember("API rate limit is 1000 req/min per workspace")
        mesh.remember("Use camelCase for JS, snake_case for Python", tags=["conventions"])
        """
        result = self._post("/remember", {
            "text":         text,
            "tags":         tags or [],
            "source_agent": source_agent,
            "confidence":   confidence,
        })
        return result["id"]

    def query(
        self,
        q: str,
        top_k: int = 5,
        tag: str = None,
        min_score: float = 0.3,
    ) -> list[dict]:
        """
        Semantic search. Returns ranked list of context entries.

        results = mesh.query("what do we know about our database?")
        for r in results:
            print(r["text"], r["score"])
        """
        result = self._post("/query", {
            "q":          q,
            "top_k":      top_k,
            "tag_filter": tag,
            "min_score":  min_score,
        })
        return result["results"]

    def forget(self, entry_id: str) -> bool:
        """
        Delete a context entry by ID.
        """
        self._delete(f"/forget/{entry_id}")
        return True

    def list(self, limit: int = 50, offset: int = 0, tag: str = None) -> dict:
        """
        Browse stored entries.
        Returns {"total": int, "entries": [...]}
        """
        params = {"limit": limit, "offset": offset}
        if tag:
            params["tag_filter"] = tag
        return self._get("/list", params)

    def usage(self) -> dict:
        """
        Check plan usage for this month.
        """
        return self._get("/usage")

    # ── Convenience ──────────────────────────────────────────────────────────

    def remember_many(self, items: list[Union[str, dict]]) -> list[str]:
        """
        Bulk store. Pass strings or dicts with 'text', 'tags', etc.
        Returns list of IDs.
        """
        ids = []
        for item in items:
            if isinstance(item, str):
                ids.append(self.remember(item))
            else:
                ids.append(self.remember(**item))
        return ids

    def __repr__(self):
        return f"Mesh(key=...{self._key[-6:]}, base={self._base})"


class AsyncMesh:
    """
    Async client for use with asyncio / FastAPI / etc.
    """

    def __init__(self, api_key: str, base_url: str = DEFAULT_BASE):
        self._key     = api_key
        self._base    = base_url.rstrip("/")
        self._headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type":  "application/json",
        }

    async def _post(self, path: str, body: dict) -> dict:
        async with httpx.AsyncClient(timeout=15.0) as c:
            r = await c.post(f"{self._base}{path}", headers=self._headers, json=body)
            self._raise(r)
            return r.json()

    async def _get(self, path: str, params: dict = None) -> dict:
        async with httpx.AsyncClient(timeout=15.0) as c:
            r = await c.get(f"{self._base}{path}", headers=self._headers, params=params or {})
            self._raise(r)
            return r.json()

    async def _delete(self, path: str) -> dict:
        async with httpx.AsyncClient(timeout=15.0) as c:
            r = await c.delete(f"{self._base}{path}", headers=self._headers)
            self._raise(r)
            return r.json()

    def _raise(self, r: httpx.Response):
        if r.status_code == 401:
            raise ContextMeshError("Invalid API key.")
        if r.status_code == 429:
            raise ContextMeshError(f"Rate/quota limit exceeded. {r.json().get('detail','')}")
        if r.status_code >= 400:
            raise ContextMeshError(f"API error {r.status_code}: {r.text}")

    async def remember(self, text: str, tags: list[str] = None, source_agent: str = None, confidence: float = 1.0) -> str:
        result = await self._post("/remember", {"text": text, "tags": tags or [], "source_agent": source_agent, "confidence": confidence})
        return result["id"]

    async def query(self, q: str, top_k: int = 5, tag: str = None, min_score: float = 0.3) -> list[dict]:
        result = await self._post("/query", {"q": q, "top_k": top_k, "tag_filter": tag, "min_score": min_score})
        return result["results"]

    async def forget(self, entry_id: str) -> bool:
        await self._delete(f"/forget/{entry_id}")
        return True

    async def list(self, limit: int = 50, offset: int = 0, tag: str = None) -> dict:
        params = {"limit": limit, "offset": offset}
        if tag:
            params["tag_filter"] = tag
        return await self._get("/list", params)

    async def usage(self) -> dict:
        return await self._get("/usage")
